### ISSUE TITLE
SpanFilter not working after upgrading to 1.0

### ISSUE THREAD

**Issue by jorgemanrubia** [DataDog/dd-trace-rb#2038]
Filtering out traces based on resource names stopped working after upgrading to the latest gem version. 

Before, with version 0.54 we had:

```ruby
Datadog::Pipeline.before_flush(
  Datadog::Pipeline::SpanFilter.new { |span| span.resource =~ /Easymon/ }
)
```

Now with 1.0 we have:

```ruby
Datadog::Tracing.before_flush(
  Datadog::Tracing::Pipeline::SpanFilter.new { |span| span.resource =~ /Easymon/ }
)
```

But it's not filtering those requests.

Is there something we should be doing differently here?

Thanks in advance.

**Comment by thauanz** [comment]
I have the same issue with a similar configuration. One thing I notice is that the filter was getting the top resource out of the trace but not the internals.
 
```
Datadog::Tracing.before_flush(
  Datadog::Tracing::Pipeline::SpanFilter.new { |span| span.resource =~ /PingController/ }
)
```

And whatever happens inside the controller would be recorded. eg: query, cache

Cheers

**Comment by marcotc** [comment]
👋 @jorgemanrubia and @thauanz, thank you for the report!
I was able to reproduce this issue. It only affects the top-level spans (say you are trying to filter out the whole trace by removing the top-most span); removing sub-spans only works correctly.

The issue seems related to our assumption documented here: https://github.com/DataDog/dd-trace-rb/blob/2703f941734253fa864d5d4e29903f86f4085f5a/lib/datadog/tracing/pipeline/span_filter.rb#L18-L20

This assumption is not new to v1.0.0, so the `SpanFilter` logic cannot be the cause of the regression alone. I suspect there are changes in other parts that might have changed the span order, which indirectly affects filtering.

Unfortunately I can't think of a temporary workaround at this moment.
I'll keep investigating and let you know.

**Comment by delner** [comment]
Span ordering changed in 1.0, particularly because of how `TraceOperation` behaves. Where the span array used to be "first-in-first-out", it's now "first-in-last-out" ordering, hence an inversion. In theory, root span is always last in 1.0. However, there are no guarantees on ordering, or that the `TraceSegment` being flushed will contain all the spans: this is because of partial flushing.

This single-pass strategy was implemented in #1071 for the sake of performance with large traces. The strategy was well-intentioned but clearly not comprehensive enough given our constraints.

In general, `SpanFilter` and `Pipeline` is a convenience feature that can be a performance liability as it puts every single span through some set of arbitrary functions. From a design perspective, I don't think `SpanFilter` should be used to eliminate entire traces, only individual spans. (It's not capable of seeing/modifying entire traces when it runs, only parts of it.)

Ideally, it shouldn't be used at all. But for a lack of a better API for customizing the output of our auto-instrumentation, we offer this.

As a workaround, if you are attempting to drop traces entirely, I would recommend adding code (like an `after_action` in Rails controllers) that checks condition and calls `trace = Datadog::Tracing.active_trace; trace.reject! if trace` if the condition is met.

Beyond that, we'll try to get some interim fix in if we can, then my sense is we'll need to address this with a bigger change down the road.

**Comment by delner** [comment]
I've merged #2074 which should restore parity of `SpanFilter` with 0.x. Please feel free to try it out, see if it works for you.
